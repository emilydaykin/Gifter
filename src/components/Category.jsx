import React from 'react';

const Category = ({ category }) => {
  const { title, imageURL, key } = category;
  return (
    <div className={`category category--${key}`}>
      <div
        className='category__image'
        style={{
          backgroundImage: `url(${imageURL})`,
          backgroundPosition: `${
            key === 'anniversary' ? 'center 60%' : key === 'thank-you' ? 'center 20%' : 'center'
          }`
        }}
      >
        <div className='category__body'>
          <h2>{title}</h2>
          <p>Shop Now</p>
        </div>
      </div>
    </div>
  );
};

export default Category;
