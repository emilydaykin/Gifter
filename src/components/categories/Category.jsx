// Displays all gifts for each category in DB
// path is `/shop/:category` for each category in DB
import { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';
import {
  selectCategoriesMap,
  selectCategoriesIsLoading
} from '../../store/categories/category.selector';
import ProductCard from '../ProductCard';
import Loader from '../Loader';
import Footer from '../Footer';

const Category = () => {
  const { category } = useParams();
  const categoriesMap = useSelector(selectCategoriesMap);
  const [products, setProducts] = useState([]);
  const isLoading = useSelector(selectCategoriesIsLoading);

  useEffect(() => {
    setProducts(categoriesMap[category]);
  }, [category, categoriesMap]);

  return (
    <>
      <div className='shop'>
        <div className='shop__section'>
          <h1 className='shop__category shop__category--individual'>{category} gifts</h1>
          {isLoading ? (
            <Loader />
          ) : (
            <div className='product-group product-group--individual'>
              {products &&
                products.map((product) => (
                  <ProductCard product={product} key={product.id} preview={false} />
                ))}
            </div>
          )}
        </div>
      </div>
      <Footer />
    </>
  );
};

export default Category;
