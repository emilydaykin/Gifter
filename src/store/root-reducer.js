import { combineReducers } from 'redux'; // to create a big final one
import { userReducer } from './user/user.reducer';
import { categoriesReducer } from './categories/category.reducer';
import { cartReducer } from './cart/cart.reducer';

export const rootReducer = combineReducers({
  user: userReducer,
  categories: categoriesReducer,
  cart: cartReducer
});
