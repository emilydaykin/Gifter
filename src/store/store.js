import { compose, createStore, applyMiddleware } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage'; // this will be by default local storage in most web browsers
import { rootReducer } from './root-reducer';
import logger from 'redux-logger';
// import { loggerMiddleware } from './middleware/logger';
import thunk from 'redux-thunk';

const persistConfig = {
  key: 'root',
  storage,
  // blacklist: ['user'] // blacklist the user reducer (since it comes from the auth state listener anyway)
  whitelist: ['cart']
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Only show logs in dev, not prod!
const middlewares = [process.env.NODE_ENV !== 'production' && logger, thunk].filter(Boolean); // enhance our store
// const middlewares = [process.env.NODE_ENV === 'development' && loggerMiddleware].filter(Boolean); // enhance our store

// Allowing Chrome to use Redux Dev Tools if Chrome extension is installed, otherwise use Redux's compose:
const composeEnhancer =
  (process.env.NODE_ENV !== 'production' &&
    window &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;

const composedEnhancers = composeEnhancer(applyMiddleware(...middlewares));

// root-reducer
export const store = createStore(persistedReducer, undefined, composedEnhancers);

export const persistor = persistStore(store);
