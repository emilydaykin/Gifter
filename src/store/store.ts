import { compose, createStore, applyMiddleware, Middleware } from 'redux';
import { persistStore, persistReducer, PersistConfig } from 'redux-persist';
import storage from 'redux-persist/lib/storage'; // this will be by default local storage in most web browsers
import { rootReducer } from './root-reducer';
import logger from 'redux-logger';
// import { loggerMiddleware } from './middleware/logger';
// import thunk from 'redux-thunk';
import createSagaMiddleware from 'redux-saga';
import { rootSaga } from './root-saga';

export type RootState = ReturnType<typeof rootReducer>;

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

type ExtendedPersistConfig = PersistConfig<RootState> & {
  whitelist: (keyof RootState)[];
};

const persistConfig: ExtendedPersistConfig = {
  key: 'root',
  storage,
  // blacklist: ['user'] // blacklist the user reducer (since it comes from the auth state listener anyway)
  whitelist: ['cart']
};

const sagaMiddleware = createSagaMiddleware();

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Only show logs in dev, not prod!
const middlewares = [process.env.NODE_ENV !== 'production' && logger, sagaMiddleware].filter(
  (middleware): middleware is Middleware => Boolean(middleware)
); // enhance our store
// const middlewares = [process.env.NODE_ENV !== 'production' && loggerMiddleware].filter(Boolean); // enhance our store

// Allowing Chrome to use Redux Dev Tools if Chrome extension is installed, otherwise use Redux's compose:
const composeEnhancer =
  (process.env.NODE_ENV !== 'production' &&
    window &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;

const composedEnhancers = composeEnhancer(applyMiddleware(...middlewares));

// root-reducer
export const store = createStore(persistedReducer, undefined, composedEnhancers);

sagaMiddleware.run(rootSaga);

export const persistor = persistStore(store);
