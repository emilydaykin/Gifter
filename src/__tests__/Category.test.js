import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Category from '../components/categories/Category';
import { Provider } from 'react-redux';
import { store } from '../store/store';

test('Category component renders', () => {
  const view = render(
    <Provider store={store}>
      <BrowserRouter>
        <Category />
      </BrowserRouter>
    </Provider>
  );

  expect.assertions(1);
  expect(view).toMatchSnapshot();
});
