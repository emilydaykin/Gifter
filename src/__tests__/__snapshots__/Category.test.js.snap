// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category component renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="shop"
      >
        <div
          class="shop__section"
        >
          <h1
            class="shop__category shop__category--individual"
          >
             gifts
          </h1>
          <div
            class="product-group product-group--individual"
          />
        </div>
      </div>
      <footer
        class="footer"
      >
        <p
          class="footer__text"
        >
          Designed & Built by
           
          <a
            class="footer__link"
            href="/about"
          >
            Emily Daykin
          </a>
           
          ~ Copyright © 2022
        </p>
      </footer>
    </div>
  </body>,
  "container": <div>
    <div
      class="shop"
    >
      <div
        class="shop__section"
      >
        <h1
          class="shop__category shop__category--individual"
        >
           gifts
        </h1>
        <div
          class="product-group product-group--individual"
        />
      </div>
    </div>
    <footer
      class="footer"
    >
      <p
        class="footer__text"
      >
        Designed & Built by
         
        <a
          class="footer__link"
          href="/about"
        >
          Emily Daykin
        </a>
         
        ~ Copyright © 2022
      </p>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
